libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -641,42 +642,24 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 		break;
 	case FunctionType::Kind::Internal:
 	{
-		optional<FunctionDefinition const*> functionDef;
-		if (auto memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
-		{
-			solUnimplementedAssert(!functionType->bound(), "Internal calls to bound functions are not yet implemented for libraries and not allowed for contracts");
+		FunctionDefinition const* unresolvedFunctionDef = IRHelpers::referencedFunctionDeclaration(_functionCall.expression());
 
-			functionDef = dynamic_cast<FunctionDefinition const*>(memberAccess->annotation().referencedDeclaration);
-			if (functionDef.value() != nullptr)
-				solAssert(functionType->declaration() == *memberAccess->annotation().referencedDeclaration, "");
-			else
-			{
-				solAssert(dynamic_cast<VariableDeclaration const*>(memberAccess->annotation().referencedDeclaration), "");
-				solAssert(!functionType->hasDeclaration(), "");
-			}
-		}
-		else if (auto identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
-		{
-			solAssert(!functionType->bound(), "");
+		solUnimplementedAssert(!memberAccess || !functionType->bound(), "");
+		solAssert(memberAccess || !functionType->bound(), "");
 
-			if (auto unresolvedFunctionDef = dynamic_cast<FunctionDefinition const*>(identifier->annotation().referencedDeclaration))
-			{
-				functionDef = &unresolvedFunctionDef->resolveVirtual(m_context.mostDerivedContract());
-				solAssert(functionType->declaration() == *identifier->annotation().referencedDeclaration, "");
-			}
-			else
-			{
-				functionDef = nullptr;
-				solAssert(dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration), "");
-				solAssert(!functionType->hasDeclaration(), "");
-			}
-		}
-		else
-			// Not a simple expression like x or A.x
-			functionDef = nullptr;
+		solAssert(!memberAccess || !unresolvedFunctionDef || functionType->declaration() == *unresolvedFunctionDef, "");
+		solAssert(unresolvedFunctionDef || !functionType->hasDeclaration(), "");
+
+		FunctionDefinition const* functionDef = (!unresolvedFunctionDef ?
+			nullptr :
+			(memberAccess ?
+				unresolvedFunctionDef :
+				&unresolvedFunctionDef->resolveVirtual(m_context.mostDerivedContract())
+			)
+		);
```

Comment:
Isn't the following easier to understand? Yes, the variable is reassigned, but that's what we have a multi-paradigm language for, don't we? I would actually even go so far and use the same variable for `unresolvedFunctionDef` and `functionDef`. After all, what does "unresolved" mean in case we do not use a virtual function call?
```suggestion
		FunctionDefinition const* functionDef = unresolvedFunctionDef;
		if (dynamic_cast<Identifier const*>(&_functionCall.expression()))
		  functionDef = unresolvedFunctionDef->resolveVirtual(m_context.mostDerivedContract());
```

Reply:
Also note that you are implictly assuming here that "no member access" and "definition found" means that we have a direct identifier.

Reply:
> Also note that you are implictly assuming here that "no member access" and "definition found" means that we have a direct identifier.

I've been hoping to keep casts to `Identifer` and `MemberAccess` hidden neatly inside `referencedFunctionDeclaration()`. Not very successfully. I'll add it back and add an assert for that case.

> Isn't the following easier to understand? Yes, the variable is reassigned, but that's what we have a multi-paradigm language for, don't we?

ok, it's indeed simpler. I shouldn't have tried so hard to get rid of `Identifier`.

> I would actually even go so far and use the same variable for `unresolvedFunctionDef` and `functionDef`. After all, what does "unresolved" mean in case we do not use a virtual function call?

In this particular case maybe reusing the variable is not that bad but in general I think it adds some unnecessary mental overhead to track what it refers to. Having separate names for things is better in my opinion. Of course it requires coming up with good names which can be hard at times.

Reply:
Done. I changed it according to your suggestions - added assertion for `identifier`, changed to reassignment and made the code reuse `functionDef` variable.

I still think that separate names would be clearer. Using the same name in assertions might give a false impression that they all still hold after the reassignment. This is not true for example for `functionType->declaration() == *functionDef`.